Custom type definition don’t get included if program contains no explicit constructor

### Environment

- Scope: desktop IDE
- Version: dev 98c13c117df904b785a26483667564e1e189c966

### Steps to reproduce

1. Create a custom type constructor `time` with a trivial alias for `Type`¹
2. Create an alternative C++ constructor `pack-cpp` for `time`: take three numbers, output time
3. Provide an implementation for the alt constructor referring the `Type`²
4. Create an example patch with `pack-cpp` on it
5. Deploy → Upload

¹
```cpp
struct State { };

// ↓↓ Important line (1) ↓↓
using Type = Number;

{{ GENERATED_CODE }}

void evaluate(Context ctx) {
    auto sec = getValue<input_SEC>(ctx);
    emitValue<output_OUT>(ctx, fmod(sec, 24 * 60 * 60));
}
```

²
```cpp
struct State { };

{{ GENERATED_CODE }}

void evaluate(Context ctx) {
    auto h = getValue<input_HOUR>(ctx);
    auto m = getValue<input_MIN>(ctx);
    auto s = getValue<input_SEC>(ctx);

    // ↓↓ Important line (2) ↓↓
    ValueType<output_OUT>::T result = h * 3600 + m * 60 + s;

    emitValue<output_OUT>(ctx, result);
}
```

Xodball for test:

[time.zip](https://github.com/xodio/xod/files/2071538/time.zip)

### Expected behavior

- The generated code compiles successfully
- The code includes `time` native implementation¹
- The native implementation comes before `pack-cpp` implementation²
- The important line (2) refers to the type defined at the important line (1)

### Actual behavior

- Compilation fails:

```
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:697:5: error: '____time' does not name a type
     ____time::Type output_OUT;
     ^
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:718:53: error: '____time' does not name a type
 template<> struct ValueType<output_OUT> { using T = ____time::Type; };
                                                     ^
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:747:12: error: '____time' does not name a type
 template<> ____time::Type getValue<output_OUT>(Context ctx) {
            ^
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:764:52: error: '____time' has not been declared
 template<> void emitValue<output_OUT>(Context ctx, ____time::Type val) {
                                                    ^
```

- The code generated indeed does not include `time`

### How to fix

If a program uses a custom type, ensure it includes the native implementation for its constructor. If it is not there push it at the head of toposorted TPatch list.

--- a/packages/xod-arduino/src/transpiler.js
+++ b/packages/xod-arduino/src/transpiler.js
@@ -98,6 +98,68 @@ const arrangeTPatchesInTopologicalOrder = def(
     )(project)
 );
 
+const convertPatchToTPatch = def(
+  'convertPatchToTPatch :: Patch -> TPatch',
+  patch => {
+    const patchPath = Project.getPatchPath(patch);
+    const impl = explodeMaybe(
+      `Implementation for ${patchPath} not found`,
+      Project.getImpl(patch)
+    );
+
+    const isDirtyable = pin =>
+      Project.getPinType(pin) === Project.PIN_TYPE.PULSE ||
+      isDirtienessEnabled(impl, `${pin.direction}_${pin.label}`);
+
+    const outputs = R.compose(
+      R.map(
+        R.applySpec({
+          type: Project.getPinType,
+          pinKey: Project.getPinLabel,
+          value: R.compose(Project.defaultValueOfType, Project.getPinType),
+          isDirtyable,
+          isDirtyOnBoot: R.compose(
+            R.not,
+            R.equals(Project.PIN_TYPE.PULSE),
+            Project.getPinType
+          ),
+        })
+      ),
+      Project.normalizePinLabels,
+      Project.listOutputPins
+    )(patch);
+
+    const inputs = R.compose(
+      R.map(
+        R.applySpec({
+          type: Project.getPinType,
+          pinKey: Project.getPinLabel,
+          isDirtyable,
+        })
+      ),
+      Project.normalizePinLabels,
+      Project.listInputPins
+    )(patch);
+
+    const isThisIsThat = {
+      isDefer: Project.isDeferNodeType(patchPath),
+      isConstant: Project.isConstantNodeType(patchPath),
+      usesTimeouts: areTimeoutsEnabled(impl),
+      usesNodeId: isNodeIdEnabled(impl),
+    };
+
+    return R.mergeAll([
+      isThisIsThat,
+      {
+        outputs,
+        inputs,
+        impl,
+        patchPath,
+      },
+    ]);
+  }
+);
+
 const createTPatches = def(
   'createTPatches :: PatchPath -> Project -> [TPatch]',
   (entryPath, project) =>
@@ -106,64 +168,8 @@ const createTPatches = def(
       // as respective nodes in a toposorted graph
       arrangeTPatchesInTopologicalOrder(entryPath, project),
       // :: Map PatchPath TPatch
-      R.mapObjIndexed((patch, patchPath) => {
-        const impl = explodeMaybe(
-          `Implementation for ${patchPath} not found`,
-          Project.getImpl(patch)
-        );
-
-        const isDirtyable = pin =>
-          Project.getPinType(pin) === Project.PIN_TYPE.PULSE ||
-          isDirtienessEnabled(impl, `${pin.direction}_${pin.label}`);
-
-        const outputs = R.compose(
-          R.map(
-            R.applySpec({
-              type: Project.getPinType,
-              pinKey: Project.getPinLabel,
-              value: R.compose(Project.defaultValueOfType, Project.getPinType),
-              isDirtyable,
-              isDirtyOnBoot: R.compose(
-                R.not,
-                R.equals(Project.PIN_TYPE.PULSE),
-                Project.getPinType
-              ),
-            })
-          ),
-          Project.normalizePinLabels,
-          Project.listOutputPins
-        )(patch);
-
-        const inputs = R.compose(
-          R.map(
-            R.applySpec({
-              type: Project.getPinType,
-              pinKey: Project.getPinLabel,
-              isDirtyable,
-            })
-          ),
-          Project.normalizePinLabels,
-          Project.listInputPins
-        )(patch);
-
-        const isThisIsThat = {
-          isDefer: Project.isDeferNodeType(patchPath),
-          isConstant: Project.isConstantNodeType(patchPath),
-          usesTimeouts: areTimeoutsEnabled(impl),
-          usesNodeId: isNodeIdEnabled(impl),
-        };
-
-        return R.mergeAll([
-          isThisIsThat,
-          {
-            outputs,
-            inputs,
-            impl,
-            patchPath,
-          },
-        ]);
-      }),
-      R.omit([entryPath]),
+      R.map(convertPatchToTPatch),
+      R.dissoc(entryPath),
       R.indexBy(Project.getPatchPath),
       Project.listGenuinePatches
     )(project)

--- a/packages/xod-arduino/src/transpiler.js
+++ b/packages/xod-arduino/src/transpiler.js
@@ -333,6 +333,28 @@ const removeUnusedNodes = def(
   }
 );
 
+const checkForNativePatchesWithTooManyOutputs = def(
+  'checkForNativePatchesWithTooManyOutputs :: TProject -> Either Error TProject',
+  tProject => {
+    const nodeWithTooManyOutputs = R.find(
+      R.pipe(R.prop('outputs'), R.length, R.lt(7)),
+      tProject.patches
+    );
+
+    if (nodeWithTooManyOutputs) {
+      return Either.Left(
+        new Error(
+          `Native node ${
+            nodeWithTooManyOutputs.patchPath
+          } has more than 7 outputs`
+        )
+      );
+    }
+
+    return Either.of(tProject);
+  }
+);
+
 /**
  * Transforms Project into TProject.
  * TProject is an object, that ready to be passed into renderer (handlebars)
@@ -342,24 +364,8 @@ const transformProjectWithImpls = def(
   'transformProjectWithImpls :: Project -> PatchPath -> TranspilationOptions -> Either Error TProject',
   (project, path, opts) =>
     R.compose(
-      R.chain(tProject => {
-        const nodeWithTooManyOutputs = R.find(
-          R.pipe(R.prop('outputs'), R.length, R.lt(7)),
-          tProject.patches
-        );
-
-        if (nodeWithTooManyOutputs) {
-          return Either.Left(
-            new Error(
-              `Native node ${
-                nodeWithTooManyOutputs.patchPath
-              } has more than 7 outputs`
-            )
-          );
-        }
-
-        return Either.of(tProject);
-      }),
+      R.chain(checkForNativePatchesWithTooManyOutputs),
+      // :: Either Error TProject
       R.map(({ project: proj, nodeIdsMap }) => {
         const patches = createTPatches(path, proj);
 

--- a/packages/xod-arduino/src/transpiler.js
+++ b/packages/xod-arduino/src/transpiler.js
@@ -51,12 +51,10 @@ const toposortProject = def(
     R.compose(
       R.chain(nodeIdsMap =>
         R.compose(
-          R.map(
-            R.applySpec({
-              project: R.identity,
-              nodeIdsMap: R.always(nodeIdsMap),
-            })
-          ),
+          R.map(transformedProject => ({
+            transformedProject,
+            nodeIdsMap,
+          })),
           () =>
             Project.updatePatch(
               path,
@@ -160,10 +158,54 @@ const convertPatchToTPatch = def(
   }
 );
 
+const getMissingExplicitConstructorTypes = def(
+  'getMissingExplicitConstructorTypes :: PatchPath -> Project -> [PatchPath]',
+  (entryPath, project) => {
+    // :: [Patch]
+    const usedPatches = R.compose(
+      // we are sure that all thote patches exist since running flatten
+      R.map(Project.getPatchByPathUnsafe(R.__, project)),
+      R.uniq,
+      R.map(Project.getNodeType),
+      Project.listNodes,
+      // we are sure that entry patch exists since running flatten
+      Project.getPatchByPathUnsafe(entryPath)
+    )(project);
+
+    // :: [PatchPath]
+    const explicitlyUsedConstructors = R.compose(
+      R.map(Project.getPatchPath),
+      R.filter(Project.isConstructorPatch)
+    )(usedPatches);
+
+    // :: [PatchPath]
+    const referencedCustomTypes = R.compose(
+      R.reject(Project.isBuiltInType),
+      R.uniq,
+      R.map(Project.getPinType),
+      R.chain(Project.listPins)
+    )(usedPatches);
+
+    return R.without(explicitlyUsedConstructors, referencedCustomTypes);
+  }
+);
+
 const createTPatches = def(
-  'createTPatches :: PatchPath -> Project -> [TPatch]',
-  (entryPath, project) =>
+  'createTPatches :: PatchPath -> Project -> Project -> [TPatch]',
+  (entryPath, project, originalProject) =>
     R.compose(
+      // Include constructor patches for custom types
+      // that are used without an explicit constructor.
+      // Otherwise, generated source code won't have
+      // type definitions for them.
+      tPatches =>
+        R.compose(
+          // Those construction patches must be at the very top
+          R.concat(R.__, tPatches),
+          R.map(convertPatchToTPatch),
+          R.map(Project.getPatchByPathUnsafe(R.__, originalProject)),
+          getMissingExplicitConstructorTypes
+        )(entryPath, project),
       // patches must appear in the same order
       // as respective nodes in a toposorted graph
       arrangeTPatchesInTopologicalOrder(entryPath, project),
@@ -372,8 +414,8 @@ const transformProjectWithImpls = def(
     R.compose(
       R.chain(checkForNativePatchesWithTooManyOutputs),
       // :: Either Error TProject
-      R.map(({ project: proj, nodeIdsMap }) => {
-        const patches = createTPatches(path, proj);
+      R.map(({ transformedProject, nodeIdsMap }) => {
+        const patches = createTPatches(path, transformedProject, project);
 
         return R.merge(
           {
@@ -382,7 +424,7 @@ const transformProjectWithImpls = def(
           R.applySpec({
             patches: R.always(patches),
             nodes: createTNodes(path, patches, nodeIdsMap),
-          })(proj)
+          })(transformedProject)
         );
       }),
       R.chain(toposortProject(path)),

--- a/packages/xod-arduino/test/fixtures/ensure-custom-types-are-defined.xodball
+++ b/packages/xod-arduino/test/fixtures/ensure-custom-types-are-defined.xodball
@@ -1,0 +1,131 @@
+{
+  "name": "",
+  "patches": {
+    "@/alternative-constructor": {
+      "attachments": [
+        {
+          "content": "struct State {\n};\n\n{{ GENERATED_CODE }}\n\nusing Type = Number;\n\nvoid evaluate(Context ctx) {\n    emitValue<output_OUT>(ctx, (Number)1337);\n}\n",
+          "encoding": "utf-8",
+          "filename": "patch.cpp"
+        }
+      ],
+      "nodes": {
+        "BJNznEEgm": {
+          "id": "BJNznEEgm",
+          "position": {
+            "x": 34,
+            "y": 102
+          },
+          "type": "@/output-my-type"
+        },
+        "H1dGnVEg7": {
+          "id": "H1dGnVEg7",
+          "position": {
+            "x": 34,
+            "y": 0
+          },
+          "type": "xod/patch-nodes/not-implemented-in-xod"
+        }
+      },
+      "path": "@/alternative-constructor"
+    },
+    "@/cast-to-string(my-type)": {
+      "attachments": [
+        {
+          "content": "#pragma XOD dirtieness disable\n\nstruct State {\n    char str[16];\n    CStringView view;\n    State() : view(str) { }\n};\n\n{{ GENERATED_CODE }}\n\nvoid evaluate(Context ctx) {\n    auto state = getState(ctx);\n    auto num = getValue<input_IN>(ctx);\n    dtostrf(num, 0, 0, state->str);\n    emitValue<output_OUT>(ctx, XString(&state->view));\n}\n",
+          "encoding": "utf-8",
+          "filename": "patch.cpp"
+        }
+      ],
+      "nodes": {
+        "SkzDTNNxQ": {
+          "id": "SkzDTNNxQ",
+          "position": {
+            "x": 34,
+            "y": 102
+          },
+          "type": "xod/patch-nodes/not-implemented-in-xod"
+        },
+        "rk7EpN4eQ": {
+          "id": "rk7EpN4eQ",
+          "position": {
+            "x": 34,
+            "y": 204
+          },
+          "type": "xod/patch-nodes/output-string"
+        },
+        "ryP7pNNlm": {
+          "id": "ryP7pNNlm",
+          "position": {
+            "x": 34,
+            "y": 0
+          },
+          "type": "@/input-my-type"
+        }
+      },
+      "path": "@/cast-to-string(my-type)"
+    },
+    "@/main": {
+      "links": {
+        "r1LqpEEe7": {
+          "id": "r1LqpEEe7",
+          "input": {
+            "nodeId": "S1BqT4VeX",
+            "pinKey": "ryP7pNNlm"
+          },
+          "output": {
+            "nodeId": "rJ97hNVxQ",
+            "pinKey": "BJNznEEgm"
+          }
+        }
+      },
+      "nodes": {
+        "S1BqT4VeX": {
+          "id": "S1BqT4VeX",
+          "position": {
+            "x": 68,
+            "y": 102
+          },
+          "type": "@/cast-to-string(my-type)"
+        },
+        "rJ97hNVxQ": {
+          "id": "rJ97hNVxQ",
+          "position": {
+            "x": 68,
+            "y": 0
+          },
+          "type": "@/alternative-constructor"
+        }
+      },
+      "path": "@/main"
+    },
+    "@/my-type": {
+      "attachments": [
+        {
+          "content": "struct State {\n};\n\nusing Type = Number;\n\n{{ GENERATED_CODE }}\n\nvoid evaluate(Context ctx) {\n    emitValue<output_OUT>(ctx, (Number)42);\n}\n",
+          "encoding": "utf-8",
+          "filename": "patch.cpp"
+        }
+      ],
+      "nodes": {
+        "Syee24Vx7": {
+          "id": "Syee24Vx7",
+          "position": {
+            "x": 68,
+            "y": 204
+          },
+          "type": "xod/patch-nodes/output-self"
+        },
+        "rJKxnE4eX": {
+          "id": "rJKxnE4eX",
+          "position": {
+            "x": 68,
+            "y": 102
+          },
+          "type": "xod/patch-nodes/not-implemented-in-xod"
+        }
+      },
+      "path": "@/my-type"
+    }
+  }
+}

--- a/packages/xod-arduino/test/transpiler.spec.js
+++ b/packages/xod-arduino/test/transpiler.spec.js
@@ -86,6 +86,29 @@ describe('xod-arduino transpiler', () => {
           'Node IDs were not arranged in topological order'
         );
       }));
+  it('ensures definitions for all used custom types are included', () =>
+    loadProject(
+      [wsPath()],
+      path.resolve(
+        __dirname,
+        './fixtures/ensure-custom-types-are-defined.xodball'
+      )
+    )
+      .then(transformProject(R.__, '@/main'))
+      .then(explodeEither)
+      .then(tProject => {
+        const actualPatchPaths = R.compose(
+          R.map(R.prop('patchPath')),
+          R.prop('patches')
+        )(tProject);
+
+        assert.deepEqual(actualPatchPaths, [
+          // implementation of "original" type constructor was added at the top
+          '@/my-type',
+          '@/alternative-constructor',
+          '@/cast-to-string(my-type)',
+        ]);
+      }));
 });
 
 describe('getNodeIdsMap', () => {

--- a/packages/xod-project/src/index.js
+++ b/packages/xod-project/src/index.js
@@ -59,6 +59,7 @@ export {
   isVariadicPatch,
   isAbstractPatch,
   isGenuinePatch,
+  isConstructorPatch,
   isPatchNotImplementedInXod,
   doesPatchHaveGenericPins,
   validateAbstractPatch,

